### DISCUSSION TITLE
Problem connecting to Clickhouse

### DISCUSSION THREAD

**Discussion by chungst** [sfu-db/connector-x#590] [Q&A]
I am trying to fetch data from Clickhouse and pipe them into Polars and I got stuck in the `dispatcher.run` part.
I have tried both `TextProtocol` and `BinaryProtocol` on `MySQLSource` but that didn't make a difference. The table is show small which only has 98 rows and it shouldn't stuck at **Start writing**  forever. Did I miss anything?

```rs
fn run() -> Result<(), Box<dyn Error>> {
    let database_url = "";
    let pool_size = 8;

    let source = MySQLSource::<TextProtocol>::new(&database_url, pool_size)?;
    let mut destination = Arrow2Destination::new();
    let queries = &["select * from tables"];
    let dispatcher = Dispatcher::<MySQLSource<TextProtocol>, Arrow2Destination, MySQLArrow2Transport<TextProtocol>>::new(source, &mut destination, queries, None);
    let r = dispatcher.run().expect("run failed");

    Ok(())
}

2024-03-26T04:35:04.285101Z DEBUG connectorx::dispatcher: Run dispatcher    
2024-03-26T04:35:04.285135Z DEBUG connectorx::dispatcher: Prepare    
2024-03-26T04:35:04.285154Z DEBUG connectorx::dispatcher: Fetching metadata    
2024-03-26T04:35:04.295518Z DEBUG connectorx::dispatcher: Do not need counts in advance    
2024-03-26T04:35:04.299392Z DEBUG connectorx::dispatcher: Allocate destination memory: 0x0    
2024-03-26T04:35:04.299445Z DEBUG connectorx::dispatcher: Create destination partition    
2024-03-26T04:35:04.299459Z DEBUG connectorx::dispatcher: Start writing    
```

**Comment by wangxiaoying** [comment]
Hi @chungst , according to the log (`Allocate destination memory: 0x0`), there is 0 columns in the result, which is pretty weird. Basically 0x0 means # row x # col. # row is 0 for arrow destination since we do not need to preallocate the buffer, but # col should not be 0. Not sure whether it is the reason that caused the issue. Can you make sure that the mysql driver can access your clickhouse?

**Reply by chungst** [reply]
Hi @wangxiaoying 
Do you mean accessing clickhouse via the MySQL/MariaDB client?
I  tried the following and I could login and issued `show tables` without problem.
```sh
mariadb  -h $HOST -u $USER -P $PORT $DB -p
```

Is there a way for me just using the source(without any destinations) to issue a SQL query? That way I could identify if the problem happened in the Source part or not.

**Reply by wangxiaoying** [reply]
> Is there a way for me just using the source(without any destinations) to issue a SQL query?

Internally connectorx uses [`mysql crate`](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L20) to connect to clickhouse (text protocol only). Here is its [repo](https://github.com/blackbeam/rust-mysql-simple) and this [query_iter](https://docs.rs/mysql/24.0.0/mysql/struct.Conn.html#method.query_iter) is the specific API we call at [here](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L263). Can you try to use this mysql crate directly and see whether if you can get any data?

**Reply by chungst** [reply]
It is strange, it connected without problem and the query result was expected and correct.
```rs
pub fn run() -> Result<(), Box<dyn Error>> {
    use mysql::*;
    use mysql::prelude::*;

    let pool = Pool::new(database_url)?;

    let mut conn = pool.get_conn()?;

    let r = conn.query_iter("SELECT * FROM branches")?.count();

    // printed 98 which is correct
    println!("{r}");

    Ok(())
}
```

**Reply by chungst** [reply]
After further digging, I found that the flow is already failing at `fetch_metadata`
[`stmt`](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L102) is not in the right shape 
```rs
[connectorx/src/sources/mysql/mod.rs:103] &stmt = Statement {
    inner: InnerStmt {
        columns: None,
        params: None,
        stmt_packet: StmtPacket {
            status: ConstU8(
                PhantomData<mysql_common::packets::InvalidStmtPacketStatus>,
            ),
            statement_id: 0,
            num_columns: 0,
            num_params: 0,
            __skip: Skip,
            warning_count: 0,
        },
        connection_id: 502,
    },
    named_params: None,
}
```
Because of [this](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L114), `src_schema` will be empty. Combining it with [1](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/destinations/arrow2/mod.rs#L60) and [2](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/dispatcher.rs#L73)
```rs
debug!(
    "Allocate destination memory: {}x{}",
    total_rows,
    src_schema.len()
);
```
It explained why **Allocate destination memory: 0x0** happened(0 rows are expected as you explained).

 However, I am not sure if it was due to the source being Clickhouse(as far as I know, clickhouse doesn't support the full MySql protocol, I could be wrong).

**Reply by wangxiaoying** [reply]
Thanks for the investigation!

This is weird, since if the stmt returns with error, our currently logic will seek to generate a query like "SELECT * from branches limit 1" to directly query the database for schema like [here](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L121). And if it still fails, it should throw an error [here](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L145) with debug and warning messages. However, it seems like there is no such messages printed in your log.

Can you please confirm the code didn't reach to [this path](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L126)?

**Reply by chungst** [reply]
The code did not reach https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L126
Given the output from `dbg!` on [these two lines.](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mysql/mod.rs#L99-L101)
```rs
[connectorx/src/sources/mysql/mod.rs:102] &first_query = Naked(
    "select * from branches",
)
[connectorx/src/sources/mysql/mod.rs:103] conn.prep(first_query) = Ok(
    Statement {
        inner: InnerStmt {
            columns: None,
            params: None,
            stmt_packet: StmtPacket {
                status: ConstU8(
                    PhantomData<mysql_common::packets::InvalidStmtPacketStatus>,
                ),
                statement_id: 0,
                num_columns: 0,
                num_params: 0,
                __skip: Skip,
                warning_count: 0,
            },
            connection_id: 563,
        },
        named_params: None,
    },
)
```
The return of `conn.prep` is `Ok` and there is no way it will reach the `Err` part. Although It looks like the problem started all the way back in the  MySQL crate(I could be wrong), I am not sure if there should be a guard internally in `connector-x` on the `Statement`

**Reply by luong-komorebi** [reply]
@chungst I wonder: did you ever get around this problem ? I am working on the same set of `dst_arrow`, `src_mysql` and I am also using `TextProtocol` as well, but i never got it working and the log also shows the same 

```
[2025-03-21T19:41:07Z DEBUG connectorx::get_arrow ] Protocol: text
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Run dispatcher
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Prepare
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Fetching metadata
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Do not need counts in advance
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Allocate destination memory: 0x0
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Create destination partition
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Start writing
[2025-03-21T19:41:07Z DEBUG connectorx::dispatcher ] Finalize partition 0
[2025-03-21T19:41:07Z ERROR <our program> ] Failed to execute query: Invalid argument error: must either specify a row count or at least one column
```

**Reply by luong-komorebi** [reply]
the query that I am running is pretty simple, just a `show table` or `select 1` or `select .... (anything)`

**Reply by chungst** [reply]
@luong-komorebi Unfortunately I didn't find a solution and I have moved on.
